src/send_kcidb.py

```diff
@@ -8,126 +8,122 @@
-            while True:
-                node = db.receive_node(sub_id)
-                self._logger.log_message(logging.INFO, f"Submitting node to \
-KCIDB: {node['_id']}")
-                sys.stdout.flush()
-
-                created_time = datetime.datetime.fromisoformat(node["created"])
-                if not created_time.tzinfo:
-                    created_time = datetime.datetime.fromtimestamp(
-                        created_time.timestamp(), tz=tz_utc)
-                revision = {
-                    "builds": [],
-                    "checkouts": [{
-                        "id": f"kernelci:{node['_id']}",
-                        "origin": "kernelci",
-                        "tree_name": node["revision"]["tree"],
-                        "git_repository_url": node["revision"]["url"],
-                        "git_commit_hash": node["revision"]["commit"],
-                        "git_repository_branch": node["revision"]["branch"],
-                        "start_time": created_time.isoformat(),
-                        "patchset_hash": "",
-                    }],
-                    "tests": [],
-                    "version": {
-                        "major": 4,
-                        "minor": 0
-                    }
-                }
-                self.send_revision(client, revision)
+    @staticmethod
+    def _validate_revision(revision):
+        return kcidb.io.SCHEMA.is_valid(revision)
 
-            sys.stdout.flush()
+    def _send_revision(self, client, revision):
+        if self._validate_revision(revision):
+            return client.submit(revision)
+        self.log.error("Aborting, invalid data")
+        sys.stdout.flush()
 
-        except KeyboardInterrupt:
-            self._logger.log_message(logging.INFO, "Stopping.")
-        except Exception:
-            self._logger.log_message(logging.ERROR, traceback.format_exc())
-        finally:
-            db.unsubscribe(sub_id)
+    def _kcidb_connect(self):
+        if not os.getenv('GOOGLE_APPLICATION_CREDENTIALS'):
+            self.log.error("GOOGLE_APPLICATION_CREDENTIALS is not set")
+            return None
+        return Client(project_id=self._kcidb_project_id,
+                      topic_name=self._kcidb_topic_name)
 
+    @staticmethod
+    def _set_timezone(created_timestamp):
+        created_time = datetime.datetime.fromisoformat(created_timestamp)
+        if not created_time.tzinfo:
+            tz_utc = datetime.timezone(datetime.timedelta(hours=0))
+            created_time = datetime.datetime.fromtimestamp(
+                created_time.timestamp(), tz=tz_utc)
+        return created_time.isoformat()
+
+    def _run(self, sub_id):
+        self.log.info("Listening for events... ")
+        self.log.info("Press Ctrl-C to stop.")
         sys.stdout.flush()
 
+        client = self._kcidb_connect()
+        if not client:
+            self.log.error("Failed to create client connection to KCIDB")
+            return False
```

Comment:
I think it would be better to do this in `self._setup()`.  Then any disconnection can be handled when an exception happens during the main loop, or the loop could be restarted without redoing the KCIDB connection etc.

Reply:
Yes, makes sense.